src/main/java/com/google/cloud/pubsublite/spark/PslSparkUtils.java

```diff
@@ -83,6 +95,67 @@ public static InternalRow toInternalRow(
     return InternalRow.apply(asScalaBufferConverter(list).asScala());
   }
 
+  @SuppressWarnings("unchecked")
+  private static <T> void extractVal(
+      StructType inputSchema,
+      InternalRow row,
+      String fieldName,
+      DataType expectedDataType,
+      Consumer<T> consumer) {
+    if (!inputSchema.getFieldIndex(fieldName).isEmpty()) {
```

Comment:
you call this twice: not sure how expensive it is? Consider saving the result instead.